Separate IONodeData type from TaskNodeData type

camielvs opened the pull request:
## Description

<!-- Please provide a brief description of the changes made in this pull request. Include any relevant context or reasoning for the changes. -->
Separates the NodeData types for IO Nodes vs Task Nodes. This allows stricter and more easily-followed type definitions for data within nodes. e.g. IO Nodes will no longer receive pointless amounts of task-node data such as `isGhost` or `callbacks`.

`taskNode` type file and `hintNode` type file have also been co-located together into a singular `nodes` type file.


No change to app functionality.

## Related Issue and Pull requests

<!-- Link to any related issues using the format #<issue-number> -->

Closes https://github.com/Shopify/oasis-frontend/issues/327

## Type of Change

<!-- Please delete options that are not relevant -->

- [x] Cleanup/Refactor

## Checklist

<!-- Please ensure the following are completed before submitting the PR -->

- [ ] I have tested this does not break current pipelines / runs functionality
- [ ] I have tested the changes on staging

## Screenshots (if applicable)

<!-- Include any screenshots that might help explain the changes or provide visual context -->

## Test Instructions

<!-- Detail steps and prerequisites for testing the changes in this PR -->
**Task Nodes**
1. Create a task node
- Drag a component from the library onto the canvas
- Verify the node renders correctly with all task information

2. Edit task node arguments
- Click on a task node
- Modify arguments in the context panel
- Verify changes are saved correctly

3. Delete a task node
- Select a task node and delete it
- Verify the node is removed from the canvas and graph

4. Duplicate a task node
- Select a task node and duplicate it
- Verify the duplicate appears with a unique ID

**Input/Output Nodes**
1. Create input and output nodes
- Add an input node to the canvas
- Add an output node to the canvas
- Verify both render with correct labels and type information

2. Edit IO node properties
- Click on an input/output node
- Modify its properties in the context panel (name, type, default value)
- Verify changes are reflected in the node

3. Connect IO nodes
- Connect an input node to a task node input
- Connect a task node output to an output node
- Verify connections work correctly

**General Operations**
1. Copy and paste nodes
- Select multiple nodes (mix of task, input, output)
- Copy and paste them
- Verify all node types paste correctly with proper data

(note some existing copy + paste functionality does not work with IO nodes)

2. Undo/redo operations
- Perform various operations on different node types
- Test undo/redo functionality
- Verify all node types maintain correct state

3. Save and reload
- Make changes to a component with various node types
- Save the component
- Reload and verify all nodes retain their data correctly

## Additional Comments

<!-- Add any additional context or information that reviewers might need to know regarding this PR -->

camielvs committed:
Stricter Typing on Nodes

Mbeaulne commented on src/components/shared/ReactFlow/FlowCanvas/IONode/IONode.tsx:
```diff
@@ -28,10 +24,12 @@ const IONode = ({ type, data, selected = false }: IONodeProps) => {
   const isOutput = type === "output";
 
-  const readOnly = !!data.readOnly;
+  const inputSpec = isInput ? (spec as InputSpec) : null;
```
Can we replace the  "unsafe" type casting with a proper type guard function. This is cleaner and more type-safe.

```ts

const isInputSpec = (
  type: "input" | "output",
  spec: InputSpec | OutputSpec,
): spec is InputSpec => {
  return type === "input";
};

...some code...

const inputSpec = isInputSpec(type, spec) ? spec : null;

...some code...
const inputValue = inputSpec?.value ?? inputSpec?.default ?? null;

```

Why its potentially better:

- No type assertions (as) - TypeScript infers the type automatically
- Self-documenting - the function name clearly expresses intent
- Type-safe - proper TypeScript type narrowing with is predicate
- Reusable - can be used elsewhere if needed
- It cleans up that confusing ternary below:

```
const inputValue = inputSpec?.value
    ? inputSpec?.value
    : inputSpec?.default
      ? inputSpec?.default
```

Mbeaulne commented on src/components/shared/ReactFlow/FlowCanvas/IONode/IONode.tsx:
```diff
@@ -28,10 +24,12 @@ const IONode = ({ type, data, selected = false }: IONodeProps) => {
+  const { spec, readOnly } = data;
+
   const isInput = type === "input";
   const isOutput = type === "output";
```
the comment below may clean up these lines as well, assuming type should always be input or output

```ts
const isInputSpec = (
  type: "input" | "output",
  spec: InputSpec | OutputSpec,
): spec is InputSpec => {
  return type === "input";
};


const isInput = isInputSpec(type, spec)


if (output && !isInput) { // this was if (output && isOutput) {
```

Mbeaulne commented on src/components/shared/ReactFlow/FlowCanvas/utils/duplicateNodes.ts:
```diff
@@ -277,17 +278,19 @@ export const duplicateNodes = (
-          [newTaskId, newTaskSpec],
-          originalNodeData,
-        );
+        const taskData = originalNode.data as TaskNodeData;
```
I think we should use type-guard functions instead of casting:

I've asked cursor to add typeguards and it came up with a pretty good solution that I think cleans up the code and makes it a little easier to read

```ts
const isTaskNode = (node: Node): node is Node<TaskNodeData> => {
  return node.type === "task";
};

const isInputNode = (node: Node): node is Node<IONodeData> => {
  return node.type === "input";
};

const isOutputNode = (node: Node): node is Node<IONodeData> => {
  return node.type === "output";
};


// Before:
const taskData = originalNode.data as TaskNodeData;
const inputData = originalNode.data as IONodeData;
const taskSpec = node.data.taskSpec as TaskSpec;

// After:
if (isTaskNode(node)) {
  const taskSpec = node.data.taskSpec; // TypeScript knows the type!
}
```
Here are some additional changes examples

![Screenshot 2025-10-17 at 12.26.48 PM.png](https://app.graphite.dev/user-attachments/assets/220a0b05-13c3-4825-b229-159194f71228.png)

camielvs replied:
Excellent suggestion

Mbeaulne commented on src/utils/nodes/createTaskNode.ts:
```diff
@@ -12,20 +12,21 @@ export const createTaskNode = (
     ...data,
     taskSpec,
     taskId,
+    connectable: connectable ?? true,
```
NIT: I don't know why, but I don't love this. can we just make connectable true by default?

```ts
const { callbacks, connectable = true, ...data } = nodeData;

const taskNodeData: TaskNodeData = {
    connectable
}
```

camielvs commented on src/components/shared/ReactFlow/FlowCanvas/FlowCanvas.tsx:
```diff
@@ -306,7 +306,7 @@ const FlowCanvas = ({
     () => ({
       connectable: !readOnly && !!nodesConnectable,
-      readOnly,
+      readOnly: !!readOnly,
```
Yes. At this point readOnly is type `boolean | undefined`. Where it is consumed it is required to be just `boolean`. That said, the smarter way about this is to give the prop a default of `false` when it enters FlowCanvas. So I've made this change.

Mbeaulne reviewed (changes requested):
[argument_editing.mov <span class="graphite__hidden">(uploaded via Graphite)</span> <img class="graphite__hidden" src="https://app.graphite.dev/user-attachments/thumbnails/e702864c-c7c1-4d95-b703-9e2fece3c50f.mov" />](https://app.graphite.dev/user-attachments/video/e702864c-c7c1-4d95-b703-9e2fece3c50f.mov)

I'm not able to update the arguments

Mbeaulne reviewed (commented):
![Screenshot 2025-10-23 at 10.25.32 AM.png](https://app.graphite.dev/user-attachments/assets/834d1d4d-aafd-4837-9eae-13816d51f15c.png)

inputs and outputs are now registering as output nodes

Mbeaulne reviewed (changes requested):
![Screenshot 2025-10-23 at 10.25.32 AM.png](https://app.graphite.dev/user-attachments/assets/834d1d4d-aafd-4837-9eae-13816d51f15c.png)

inputs and outputs are now registering as output nodes

camielvs commented:
Fixed - it appears we don't have enough info for the new `isInputSpec` method to infer whether a node is an Input or Output. I've now added a secondary check via `node.type`.

maxy-shpfy reviewed (commented):
Small regression (compared to staging - at the end of the video):

1. create Input node. It is marked as "INVALID" at 3 places (pipeline validation, canvas node, context panel).
2. input value. Validation now passed.
3. remove value. Validation fails again. But now the node is Marked as INVALID just in 2 places (canvas node skipped). --- BUG

[Screen Recording 2025-10-23 at 4.24.47 PM.mov <span class="graphite__hidden">(uploaded via Graphite)</span> <img class="graphite__hidden" src="https://app.graphite.dev/user-attachments/thumbnails/ffc9d237-d24f-44cd-89a7-49535acdc4ae.mov" />](https://app.graphite.dev/user-attachments/video/ffc9d237-d24f-44cd-89a7-49535acdc4ae.mov)

camielvs commented:
Fixed!

camielvs commented:
Closed in favour of v2 Editor migration. See https://github.com/TangleML/tangle-ui/pull/2028
